Sundtekgrabber

niklasstich opened the pull request:
Implemented both Crawler and Grabber for the Sundtek Media Server component
Changed ContentInformation and interface for IGrabber as discussed in the meeting

niklasstich committed:
Added Data and MediaInput.Test projects
Wrote M3UParser that can parse the specific .m3u files the sundtek server sends
Minor changes to ContentInformation and IGrabber

niklasstich committed:
added crawler for sundtek stick and started work on it

niklasstich committed:
finished implementing MySQL logic for sundtek content

niklasstich committed:
implemented and tested GetMediaStream for sundtek

AlexanderWalk reviewed (approved):
Sollt passe

Edit: 
Bei `GetAvailableContentInformation()` und `GetAvailableContentInformation(string category)`sollte der Aufruf umgedreht werden: Die Query sollte pro Kategorie in `GetAvailableContentInformation(string category)` stattfinden und `GetAvailableContentInformation()` ruft die andere Methode `foreach string` von `GetAvailableCategories` auf. 

Wäre glaub ich schöner. 

Edit2: Dafür müsste jedoch die Datenbank vermutlich für jede Kategorie einmal durchiteriert werden - hinfällig.

niklasstich commented:
> Sollt passe
> 
> Edit:
> Bei `GetAvailableContentInformation()` und `GetAvailableContentInformation(string category)`sollte der Aufruf umgedreht werden: Die Query sollte pro Kategorie in `GetAvailableContentInformation(string category)` stattfinden und `GetAvailableContentInformation()` ruft die andere Methode `foreach string` von `GetAvailableCategories` auf.
> 
> Wäre glaub ich schöner.
> 
> Edit2: Dafür müsste jedoch die Datenbank vermutlich für jede Kategorie einmal durchiteriert werden - hinfällig.

Das sollte schon gehen, man sollte halt erst schauen ob die Kategorie vorhanden ist in AvailableCategories, und dann ändert man das SelectCommand zu `SELECT * FROM sundtekcontent WHERE Category = {category}`
Ich glaube zwar, dass das performancetechnisch schlechter ist, aber wäre auf jeden Fall eleganterer Code

niklasstich committed:
Reworked the way we get data from the sql server, we now get categories seperately and them put them together to form a full collection
Changed IGrabber interface to return a dictionary with key of string and value of IEnumerable<ContentInformation>

niklasstich commented on MediaInput/IGrabber.cs:
```diff
@@ -16,8 +15,8 @@ public interface IGrabber
         /// </summary>
-        /// <returns>A two-dimensional list of all the categories and containing content this <c>IGrabber</c> has access to.</returns>
-        public IEnumerable<IEnumerable<ContentInformation>> GetAvailableContentInformation();
+        /// <returns>A dictionary containing the category and a list of content this <c>IGrabber</c> has access to.</returns>
+        public IDictionary<string, IEnumerable<ContentInformation>> GetAvailableContentInformation();
```
Wir geben jetzt ein IDictionary zurück @AlexanderWalk

niklasstich commented on MediaInput/SundtekGrabber.cs:
```diff
@@ -1,52 +1,82 @@
+        public IDictionary<string, IEnumerable<ContentInformation>> GetAvailableContentInformation()
         {
-            //TODO: Implement this by getting the playlist files from the sundtek server and "parsing" the .m3u files
-            throw new NotImplementedException();
+            return GetAvailableCategories().ToDictionary(category => category, GetAvailableContentInformation);
         }
 
         public IEnumerable<ContentInformation> GetAvailableContentInformation(string category)
         {
-            throw new NotImplementedException();
+            
+            using (var conn =
+                new MySqlConnection(
+                    $"Server={_config["MySqlServerAddress"]};Database={_config["MySqlServerDatabase"]};Uid={_config["MySqlServerUser"]};Pwd={_config["MySqlServerPassword"]};")
+            )
+            {
+                var selectCommand = new MySqlCommand($"SELECT * FROM sundtekcontent WHERE Category=@category", conn);
+                selectCommand.Parameters.AddWithValue("@category", category);
+                var adapter = new MySqlDataAdapter
+                    {SelectCommand = selectCommand};
+                var dataset = new DataSet();
+
+                adapter.Fill(dataset);
+
+                var rowCollection = dataset.Tables[0].Rows;
+
+                return (from DataRow entry in rowCollection select new ContentInformation((string) entry[0], (string) entry[1], (string) entry[2], Convert.ToBoolean(entry[3]), Convert.ToBoolean(entry[4]), new Uri((string) entry[5]), new Uri((string) entry[6]))).ToList();
+            }
```
Hier holen wir jetzt jede Category seperat und bauen sie dann zu einem Dictionary mit dem Key zusammen

DominikWiesner reviewed (approved):
Approved!

michaelpezke reviewed (approved):
Confirmed!
